pkgs/build-support/node/fetch-pnpm-deps/default.nix

```
@@ -0,0 +1,101 @@
+{ lib
+, stdenvNoCC
+, fetchurl
+, nodePackages
+, jq
+, moreutils
+, cacert
+, makeSetupHook
+}: let
+  # The names of these attributes refer to the pnpm lockfile version
+  pinnedVersions = {
+    "6_0" = {
+      name = "pnpm-8.15.5";
+      src = fetchurl {
+        url = "https://registry.npmjs.org/pnpm/-/pnpm-8.15.5.tgz";
+        sha512 = "sFGjLH5pWDO4SSbTspuMylclS1ifBknYmcbp0O22cLkex+KkNFm65zdZu1zmGcMmbxFr+THOItHvF1mn5Fqpbw==";
+      };
+    };
+    "7_0" = {
+      name = "pnpm-9.0.6";
+      src = fetchurl {
+        url = "https://registry.npmjs.org/pnpm/-/pnpm-9.0.6.tgz";
+        hash = "sha256-BiTjDv+GbN6zY7FQYb23/ZQlsXvBu0LCL19O/eoh9rM=";
+      };
+    };
+  };
+
+  pnpms = builtins.mapAttrs (_: value: nodePackages.pnpm.overrideAttrs value) pinnedVersions;
+
+  mkFetchPnpmDeps = version: let
+    pnpm = pnpms.${version};
+  in {
+    inherit pnpm;
+
+    fetchPnpmDeps = { src
+      , hash ? ""
+      , pname
+      , ...
+      } @ args:
+      let
+        args' = builtins.removeAttrs args [ "hash" "pname" "pnpm" ];
+        hash' =
+          if hash != "" then {
+            outputHash = hash;
+          } else {
+            outputHash = "";
+            outputHashAlgo = "sha256";
+          };
+      in
+      stdenvNoCC.mkDerivation (args' // {
+        name = "${pname}-pnpm-${version}-deps";
+
+        nativeBuildInputs = [
+          jq
+          moreutils
+          pnpm
+          cacert
+        ];
+
+        # https://github.com/NixOS/nixpkgs/blob/763e59ffedb5c25774387bf99bc725df5df82d10/pkgs/applications/misc/pot/default.nix#L56
+        installPhase = ''
+          runHook preInstall
+
+          export HOME=$(mktemp -d)
+          pnpm config set store-dir $out
+          # pnpm is going to warn us about using --force
+          # --force allows us to fetch all dependencies including ones that aren't meant for our host platform
+          pnpm install --frozen-lockfile --ignore-script --force
+
+          runHook postInstall
+        '';
+
+        fixupPhase = ''
+          runHook preFixup
+
+          rm -rf $out/v3/tmp
+          for f in $(find $out -name "*.json"); do
+            sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+            jq --sort-keys . $f | sponge $f
+          done
```

```suggestion
          # Remove timestamp and sort the json files
          rm -rf $out/v3/tmp
          for f in $(find $out -name "*.json"); do
            sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
            jq --sort-keys . $f | sponge $f
          done
```
we should describe what this is doing. i used the comment from `pot`